#### Mohamed Abdelwahab 43-18031

#### Regression Assignment - Deadline & Evaluation Sunday 12 June

#### You are required to use the below dataset on house prices to train a regression model:

#### https://www.kaggle.com/datasets/vikrishnan/boston-house-prices

#### Requirements:

#### - Divide dataset after importing into Train/Cross Validation/Test

#### - Test various hypothesis with different degrees and optimize your model and parameters using cross validation & test sets as explained in the lecture

#### - Apply Regularization and optimize the value of its parameter -->

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

from sklearn.datasets import load_boston

from sklearn.preprocessing import StandardScaler, MinMaxScaler, MaxAbsScaler, Normalizer, RobustScaler
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import Ridge
from sklearn.linear_model import RidgeCV
from sklearn.linear_model import LassoCV

from sklearn.pipeline import make_pipeline

from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV

from sklearn.metrics import mean_absolute_error, mean_squared_error
from sklearn.metrics import r2_score
from sklearn import metrics

import warnings
warnings.filterwarnings('ignore')

In [2]:
data = load_boston()
X = data.data
Y = data.target
X.shape, Y.shape

((506, 13), (506,))

In [3]:
df_x = pd.DataFrame(X)
df_y = pd.DataFrame(Y)
df_x

0     1      2    3      4      5     6       7    8      9     10  \
0    0.00632  18.0   2.31  0.0  0.538  6.575  65.2  4.0900  1.0  296.0  15.3   
1    0.02731   0.0   7.07  0.0  0.469  6.421  78.9  4.9671  2.0  242.0  17.8   
2    0.02729   0.0   7.07  0.0  0.469  7.185  61.1  4.9671  2.0  242.0  17.8   
3    0.03237   0.0   2.18  0.0  0.458  6.998  45.8  6.0622  3.0  222.0  18.7   
4    0.06905   0.0   2.18  0.0  0.458  7.147  54.2  6.0622  3.0  222.0  18.7   
..       ...   ...    ...  ...    ...    ...   ...     ...  ...    ...   ...   
501  0.06263   0.0  11.93  0.0  0.573  6.593  69.1  2.4786  1.0  273.0  21.0   
502  0.04527   0.0  11.93  0.0  0.573  6.120  76.7  2.2875  1.0  273.0  21.0   
503  0.06076   0.0  11.93  0.0  0.573  6.976  91.0  2.1675  1.0  273.0  21.0   
504  0.10959   0.0  11.93  0.0  0.573  6.794  89.3  2.3889  1.0  273.0  21.0   
505  0.04741   0.0  11.93  0.0  0.573  6.030  80.8  2.5050  1.0  273.0  21.0   

         11    12  
0    396.90  4.98  
1    396.90  9.14  
2    392.83  4.03  
3    394.63  2.94  
4    396.90  5.33  
..      ...   ...  
501  391.99  9.67  
502  396.90  9.08  
503  396.90  5.64  
504  393.45  6.48  
505  396.90  7.88  

[506 rows x 13 columns]

In [4]:
x_train, x_test, y_train, y_test = train_test_split(df_x, df_y, test_size=0.1, random_state=101)
x_train.shape, x_test.shape

((455, 13), (51, 13))

In [5]:
def polynomial_features(degree, train, test):
    
    poly = PolynomialFeatures(degree=degree, include_bias=False)
    train_poly = poly.fit_transform(train)
    test_poly = poly.transform(test)
    
    return train_poly, test_poly

In [6]:
def polynomial_regression(degree, cv, train, test):
    
    train_poly, test_poly = polynomial_features(degree, train, test)
    
    
    print('Train data and test data shapes are: ', train_poly.shape, test_poly.shape)
    
    model = LinearRegression(normalize=True)
    param_grid = {'fit_intercept': [True]}

    print('----------------------------------')
    print('degree = ', degree)
    print('CV = ', cv)
    print('----------------------------------')
    
    model_gscv = GridSearchCV(model, param_grid = param_grid, cv=cv, 
                              verbose=3, scoring='neg_mean_absolute_error')
    model_gscv.fit(train_poly, y_train)
    
    print('----------------------------------')
    scores = []
    best_cv = 0
    for i in range(cv):

        scores.append(model_gscv.cv_results_['split'+str(i)+'_test_score'][0])
        best_cv = scores.index(max(scores))
    best_score = -max(scores)
    print('mean_score = ', -np.mean(scores))
    print('best_score = ', best_score)
    print('best_cv = ', best_cv+1)
    print('----------------------------------')
    
    print('*******************************************')
    return 

In [7]:
for i in range(1,11):
    
    polynomial_regression(i, 5, x_train, x_test)

Train data and test data shapes are:  (455, 13) (51, 13)
----------------------------------
degree =  1
CV =  5
----------------------------------
Fitting 5 folds for each of 1 candidates, totalling 5 fits
[CV 1/5] END ...............fit_intercept=True;, score=-3.840 total time=   0.0s
[CV 2/5] END ...............fit_intercept=True;, score=-3.524 total time=   0.0s
[CV 3/5] END ...............fit_intercept=True;, score=-2.559 total time=   0.0s
[CV 4/5] END ...............fit_intercept=True;, score=-3.330 total time=   0.0s
[CV 5/5] END ...............fit_intercept=True;, score=-3.217 total time=   0.0s
----------------------------------
mean_score =  3.2938781636863226
best_score =  2.5592886483883994
best_cv =  3
----------------------------------
*******************************************
Train data and test data shapes are:  (455, 104) (51, 104)
----------------------------------
degree =  2
CV =  5
----------------------------------
Fitting 5 folds for each of 1 candidates, total

##### As it appears in the previous cell, polynomial regression with degree = 2 gives the best performance.
##### The best score at each degree is:
##### degree 1: 2.559
##### degree 2: 1.982
##### degree 3: 77.729
##### degree 4: 29.312
##### degree 5: 27.692
##### degree 6: 29.805
##### degree 7: 29.834
##### degree 8: 36.555
##### degree 9: 46.673
##### degree 10: 62.220

##### Note: increasing the degree results in remarkable inrease in total time per fit
##### Note: increasing the degeree results in huge increase in number of features
##### Also another note, always CV 3/5 and 4/5 give the best results.

### Now, trying different n_splits (cv) and reducing the maximum degree

In [8]:
for i in range(1,5):
    
    for cv in range(5,11):
    
        polynomial_regression(i, cv, x_train, x_test)

Train data and test data shapes are:  (455, 13) (51, 13)
----------------------------------
degree =  1
CV =  5
----------------------------------
Fitting 5 folds for each of 1 candidates, totalling 5 fits
[CV 1/5] END ...............fit_intercept=True;, score=-3.840 total time=   0.0s
[CV 2/5] END ...............fit_intercept=True;, score=-3.524 total time=   0.0s
[CV 3/5] END ...............fit_intercept=True;, score=-2.559 total time=   0.0s
[CV 4/5] END ...............fit_intercept=True;, score=-3.330 total time=   0.0s
[CV 5/5] END ...............fit_intercept=True;, score=-3.217 total time=   0.0s
----------------------------------
mean_score =  3.2938781636863226
best_score =  2.5592886483883994
best_cv =  3
----------------------------------
*******************************************
Train data and test data shapes are:  (455, 13) (51, 13)
----------------------------------
degree =  1
CV =  6
----------------------------------
Fitting 6 folds for each of 1 candidates, totalli

[CV 5/7] END ...............fit_intercept=True;, score=-3.066 total time=   0.0s
[CV 6/7] END ...............fit_intercept=True;, score=-3.049 total time=   0.0s
[CV 7/7] END ...............fit_intercept=True;, score=-2.345 total time=   0.0s
----------------------------------
mean_score =  2.5995228669198895
best_score =  1.923890203406105
best_cv =  4
----------------------------------
*******************************************
Train data and test data shapes are:  (455, 104) (51, 104)
----------------------------------
degree =  2
CV =  8
----------------------------------
Fitting 8 folds for each of 1 candidates, totalling 8 fits
[CV 1/8] END ...............fit_intercept=True;, score=-2.834 total time=   0.0s
[CV 2/8] END ...............fit_intercept=True;, score=-2.087 total time=   0.0s
[CV 3/8] END ...............fit_intercept=True;, score=-3.363 total time=   0.0s
[CV 4/8] END ...............fit_intercept=True;, score=-1.705 total time=   0.0s
[CV 5/8] END ...............fit_i

[CV 4/10] END ............fit_intercept=True;, score=-142.898 total time=   0.0s
[CV 5/10] END .............fit_intercept=True;, score=-59.094 total time=   0.0s
[CV 6/10] END .............fit_intercept=True;, score=-80.671 total time=   0.0s
[CV 7/10] END ............fit_intercept=True;, score=-181.039 total time=   0.0s
[CV 8/10] END .............fit_intercept=True;, score=-51.454 total time=   0.0s
[CV 9/10] END .............fit_intercept=True;, score=-98.415 total time=   0.0s
[CV 10/10] END ............fit_intercept=True;, score=-93.489 total time=   0.0s
----------------------------------
mean_score =  103.79197064294513
best_score =  47.448604297160635
best_cv =  3
----------------------------------
*******************************************
Train data and test data shapes are:  (455, 2379) (51, 2379)
----------------------------------
degree =  4
CV =  5
----------------------------------
Fitting 5 folds for each of 1 candidates, totalling 5 fits
[CV 1/5] END ..............fit

#### Best score overall is = 1.5939 at CV(5/10), degree = 2
#### Now, lets evaluate our model on the test data

In [9]:
x_train_poly, x_test_poly = polynomial_features(2 , x_train, x_test)
print('Train data and test data shapes are: ', x_train_poly.shape, x_test_poly.shape)

Train data and test data shapes are:  (455, 104) (51, 104)


#### Let's try to slice the data where CV(5/10) with a step = 46

In [10]:
x_train_poly_slice  = np.delete(x_train_poly, [*range(228,274)], 0)
y_train_slice  = np.delete(np.array(y_train), [*range(228,274)], 0)

In [11]:
model = LinearRegression().fit(x_train_poly_slice, y_train_slice)

y_pred_train = model.predict(x_train_poly_slice)
y_pred_test = model.predict(x_test_poly)

print('train data mean_absolute_error score = ', mean_absolute_error(y_pred_train, y_train_slice))
print('test data mean_absolute_error score = ', mean_absolute_error(y_pred_test, y_test))
print("\n")

print("R-Square Value",r2_score(y_test,y_pred_test))
print("\n")
print ("mean_squared_error : ",metrics.mean_squared_error(y_test, y_pred_test))
print("\n")
print ("root_mean_squared_error : ",np.sqrt(metrics.mean_squared_error(y_test, y_pred_test)))
print("\n")

print ("iNTERCEPT : ",model.intercept_)
print ("CO-EFFICIENT : ",model.coef_)

train data mean_absolute_error score =  1.735068608482039
test data mean_absolute_error score =  2.3865343864347928


R-Square Value 0.9251833853452617


mean_squared_error :  9.771911084767247


root_mean_squared_error :  3.126005611761957


iNTERCEPT :  [-89.7367187]
CO-EFFICIENT :  [[-6.33889006e+00  3.52890682e-01 -4.21176506e+00  1.72281762e+01
   9.18193598e+00  1.61077757e+01  9.85121036e-01 -1.27085235e+01
   2.81825307e+00 -4.61270456e-02  4.67104198e+00  6.26488614e-02
   1.08098393e+00  9.86816819e-04  1.16434639e-01  1.71296989e-01
   1.95930844e+00 -8.45230952e-01  1.44175546e-01 -1.99669950e-03
  -7.19302881e-02 -8.89089545e-02 -2.01178908e-03  3.02869851e-01
  -2.73335507e-04  2.18880568e-02 -5.09246024e-04 -8.76756553e-03
  -3.52316061e-02 -1.03369120e+00 -3.91131316e-03 -6.14312115e-04
  -1.41037652e-02  1.26070776e-03  6.10022717e-04 -1.04774793e-02
   8.27519755e-04 -3.04786865e-03  2.57869568e-02 -5.77939151e-02
   1.23719121e+00  3.50522348e-01  4.11181458e-04  8.3

In [12]:
pd.DataFrame(np.concatenate((y_pred_train,y_train_slice), axis=1), columns=['Predicted', 'True'])

Predicted  True
0    24.497354  21.9
1    19.592904  17.4
2    16.725480  15.0
3    33.156314  34.7
4    13.625101  13.8
..         ...   ...
404  23.702619  25.0
405  24.788903  23.0
406  19.300875  18.5
407  20.068601  18.9
408  24.752821  24.1

[409 rows x 2 columns]

In [13]:
pd.DataFrame(np.concatenate((y_pred_test,y_test), axis=1), columns=['Predicted', 'True'])

Predicted  True
0   47.880668  50.0
1   31.840582  36.2
2   14.666237  11.7
3   15.640750  14.9
4   29.005990  30.8
5   33.060292  36.1
6   46.566167  50.0
7   16.444430  13.2
8   35.329882  41.3
9    7.349907   8.5
10  27.472798  23.5
11  13.103184  13.3
12  13.168349  20.2
13  25.638321  18.6
14  22.551942  18.5
15  26.199126  27.9
16   9.090573  10.4
17  31.114001  28.5
18  27.854790  28.4
19  24.828469  23.1
20  10.744695  12.8
21  24.094169  21.7
22  15.971846  11.9
23  29.659676  29.6
24  34.153380  37.3
25  18.486060  19.9
26  27.594438  29.0
27  18.653489  18.0
28  40.378482  50.0
29  33.167255  34.6
30  20.952829  24.3
31  17.957765  18.2
32  37.228929  44.0
33  51.473338  50.0
34  29.935849  30.7
35  22.604476  22.6
36  13.811414  14.1
37  16.454426  14.5
38   5.646214   8.4
39  31.525577  32.5
40  21.616403  24.3
41  21.170362  20.6
42  41.829660  43.8
43  15.049194  11.5
44  19.992781  19.9
45  22.817022  23.8
46  31.354908  28.7
47  16.483417  17.8
48  24.737993  23.9
49  31.849857  33.0
50  33.568402  31.5

#### Now using the entire training dataset

In [14]:
model = LinearRegression().fit(x_train_poly, y_train)

y_pred_train = model.predict(x_train_poly)
y_pred_test = model.predict(x_test_poly)

print('train data score = ', mean_absolute_error(y_pred_train, y_train))
print('test data score = ', mean_absolute_error(y_pred_test, y_test))
print("\n")

print("R-Square Value",r2_score(y_test,y_pred_test))
print("\n")
print ("mean_squared_error : ",metrics.mean_squared_error(y_test, y_pred_test))
print("\n")
print ("root_mean_squared_error : ",np.sqrt(metrics.mean_squared_error(y_test, y_pred_test)))
print("\n")

print ("iNTERCEPT : ",model.intercept_)
print ("CO-EFFICIENT : ",model.coef_)

train data score =  1.7504530500336486
test data score =  2.5615224431224837


R-Square Value 0.9169900070751318


mean_squared_error :  10.842060600473829


root_mean_squared_error :  3.2927284431719888


iNTERCEPT :  [-194.35760515]
CO-EFFICIENT :  [[-5.24301283e+00  2.73768321e-01 -4.37952953e+00  2.08342543e+01
   1.69044342e+02  2.45812487e+01  9.42750069e-01 -8.31765649e+00
   2.47771511e+00 -2.74705822e-02  7.17947176e+00  9.25086583e-02
   8.53478689e-01  1.24972866e-03  6.22368047e-02  2.10710761e-01
   1.82230549e+00 -3.36778101e-01  1.45693752e-01 -2.74163980e-03
  -6.32082847e-02 -3.39810854e-02 -5.70223447e-03  2.56076959e-01
  -2.84196668e-04  1.96675994e-02 -6.81689552e-04 -9.46337189e-03
  -1.52279604e-02 -1.17645313e+00  2.87451841e-03 -2.21480678e-04
  -1.54644095e-02 -2.71737115e-03  6.29691569e-04 -7.57174210e-03
   9.82781038e-04 -2.84954581e-03  2.73826410e-02  3.41755804e-02
   1.03891536e+00  2.62262729e-01  2.30750047e-03  1.11548657e-01
  -1.99129554e-02  4.66

In [15]:
pd.DataFrame(np.concatenate((y_pred_train,y_train), axis=1), columns=['Predicted', 'True'])

Predicted  True
0    25.291131  21.9
1    18.917234  17.4
2    16.329523  15.0
3    32.537481  34.7
4    13.398811  13.8
..         ...   ...
450  23.436057  25.0
451  24.914311  23.0
452  19.468876  18.5
453  20.575015  18.9
454  25.787463  24.1

[455 rows x 2 columns]

In [16]:
pd.DataFrame(np.concatenate((y_pred_test,y_test), axis=1), columns=['Predicted', 'True'])

Predicted  True
0   48.524532  50.0
1   31.306076  36.2
2   16.029514  11.7
3   16.253900  14.9
4   28.842445  30.8
5   33.213107  36.1
6   46.170114  50.0
7   16.750013  13.2
8   35.816179  41.3
9    7.258280   8.5
10  26.594667  23.5
11  13.028115  13.3
12  13.375488  20.2
13  26.321707  18.6
14  22.420442  18.5
15  25.453454  27.9
16   9.094386  10.4
17  31.176918  28.5
18  27.169833  28.4
19  24.788678  23.1
20  10.550061  12.8
21  24.471279  21.7
22  16.852640  11.9
23  30.439481  29.6
24  34.259886  37.3
25  17.938789  19.9
26  27.898734  29.0
27  18.135937  18.0
28  40.161147  50.0
29  33.405971  34.6
30  21.009654  24.3
31  17.984329  18.2
32  37.129573  44.0
33  54.515201  50.0
34  30.108840  30.7
35  22.521917  22.6
36  13.978037  14.1
37  16.144526  14.5
38   4.994672   8.4
39  31.305268  32.5
40  21.557048  24.3
41  20.976149  20.6
42  41.783405  43.8
43  14.800237  11.5
44  19.549588  19.9
45  22.606426  23.8
46  30.899458  28.7
47  16.131573  17.8
48  24.703071  23.9
49  31.247663  33.0
50  33.538420  31.5

#### so, using the sliced data gives better results, now lets explore regularization in regression techniques

#### Ridge regression

#### alpha: must be positive, increase for more regularization

In [17]:
def range_with_floats(start, stop, step):
    while stop > start:
        yield start
        start += step

In [18]:
for a in range_with_floats(0, 0.1, 0.01):
    
    ridgereg = Ridge(alpha=a, normalize=True)
    ridgereg.fit(x_train_poly_slice, y_train_slice)

    y_pred_train = ridgereg.predict(x_train_poly_slice)
    y_pred_test = ridgereg.predict(x_test_poly)
    
    print('alpha = ', a)
    print('-----------------------')
    print('train data score = ', mean_absolute_error(y_pred_train, y_train_slice))
    print('test data score = ', mean_absolute_error(y_pred_test, y_test))
    print('-----------------------')
    
    print("R-Square Value",r2_score(y_test,y_pred_test))
    print("\n")
    print ("mean_squared_error : ",metrics.mean_squared_error(y_test, y_pred_test))
    print("\n")
    print ("root_mean_squared_error : ",np.sqrt(metrics.mean_squared_error(y_test, y_pred_test)))
    print("\n")
    
    print('-----------------------')
    print('*******************************************')

# print ("iNTERCEPT : ",ridgereg.intercept_)
# print ("CO-EFFICIENT : ",ridgereg.coef_)

alpha =  0
-----------------------
train data score =  1.7350686084882259
test data score =  2.3865343865296036
-----------------------
R-Square Value 0.9251833853400826


mean_squared_error :  9.771911085443696


root_mean_squared_error :  3.126005611870154


-----------------------
*******************************************
alpha =  0.01
-----------------------
train data score =  2.097342661270156
test data score =  2.626621783484253
-----------------------
R-Square Value 0.8967255184739821


mean_squared_error :  13.48883607544745


root_mean_squared_error :  3.672715082258281


-----------------------
*******************************************
alpha =  0.02
-----------------------
train data score =  2.2100793015837654
test data score =  2.726048065041149
-----------------------
R-Square Value 0.8858190542623745


mean_squared_error :  14.913345845327498


root_mean_squared_error :  3.8617801394340794


-----------------------
*******************************************
alpha = 

In [19]:
# create an array of alpha values
alpha_range = [*range_with_floats(0.01, 0.1, 0.01)]

# select the best alpha with RidgeCV
ridgeregcv = RidgeCV(alphas=alpha_range, normalize=True, scoring='neg_mean_absolute_error')
ridgeregcv.fit(x_train_poly_slice, y_train_slice)
print('ridgeregcv.alpha = ' , ridgeregcv.alpha_)


# predict method uses the best alpha value
y_pred_train = ridgeregcv.predict(x_train_poly_slice)
y_pred_test = ridgeregcv.predict(x_test_poly)
# calculate R^2 value, MAE, MSE, RMSE

print('-----------------------')
print('train data score = ', mean_absolute_error(y_pred_train, y_train_slice))
print('test data score = ', mean_absolute_error(y_pred_test, y_test))
print('-----------------------')

print("R-Square Value",r2_score(y_test,y_pred_test))
print("\n")
print ("mean_absolute_error :",metrics.mean_absolute_error(y_test, y_pred_test))
print("\n")
print ("mean_squared_error : ",metrics.mean_squared_error(y_test, y_pred_test))
print("\n")
print ("root_mean_squared_error : ",np.sqrt(metrics.mean_squared_error(y_test, y_pred_test)))

ridgeregcv.alpha =  0.01
-----------------------
train data score =  2.0973426612701482
test data score =  2.6266217834858576
-----------------------
R-Square Value 0.8967255184738361


mean_absolute_error : 2.6266217834858576


mean_squared_error :  13.488836075466532


root_mean_squared_error :  3.6727150822608787


#### As we can see, increasing alpha (increasing regularization), increases the error
#### Let's try 
#### Lasso regression
#### alpha: must be positive, increase for more regularization

In [20]:
# select the best alpha with RidgeCV
lassoregcv = LassoCV(n_alphas=100, normalize=True, random_state=1)
lassoregcv.fit(x_train_poly, y_train)
print('alpha : ',lassoregcv.alpha_)

# predict method uses the best alpha value
y_pred_train = lassoregcv.predict(x_train_poly)
y_pred_test = lassoregcv.predict(x_test_poly)
# calculate R^2 value, MAE, MSE, RMSE

print('-----------------------')
print('train data score = ', mean_absolute_error(y_pred_train, y_train))
print('test data score = ', mean_absolute_error(y_pred_test, y_test))
print('-----------------------')

print("R-Square Value",r2_score(y_test,y_pred_test))
print("\n")
print ("mean_absolute_error :",metrics.mean_absolute_error(y_test, y_pred_test))
print("\n")
print ("mean_squared_error : ",metrics.mean_squared_error(y_test, y_pred_test))
print("\n")
print ("root_mean_squared_error : ",np.sqrt(metrics.mean_squared_error(y_test, y_pred_test)))

alpha :  0.00031092792328413373
-----------------------
train data score =  1.9528597417819653
test data score =  2.763634646192633
-----------------------
R-Square Value 0.8962169968242343


mean_absolute_error : 2.763634646192633


mean_squared_error :  13.55525485647553


root_mean_squared_error :  3.6817461694793043
